### ISSUE TITLE
Multiple training runs overwriting labels.cache

### ISSUE THREAD

**Issue by rnguyen-bah** [ultralytics/ultralytics#21126]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi, I'm training multiple models at once using variations of dataset combinations on an HPC system with slurm. I'm wondering if I submit multiple training jobs and they are executed at the same time will there be issues if the labels.cache file is overwritten?

Here are 2 training configs for example,

**_training_config1.yaml_**
```
train:
  - /path/to/dataset1/train/images
val:
  - /path/to/dataset1/val/images

names:
  0: class0
  1: class1
  2: class2
```

**_training_config2.yaml_**
```
train:
  - /path/to/dataset1/train/images
  - /path/to/dataset2/train/images
val:
  - /path/to/dataset1/val/images

names:
  0: class0
  1: class1
  2: class2
```

And I submit 2 jobs with slurm using the following training commands:
- Run 1) `yolo detect train data=training_config1.yaml model=yolo11n.yaml epochs=100 imgsz=640`
- Run 2) `yolo detect train data=training_config2.yaml model=yolo11n.yaml epochs=100 imgsz=640`

Since the first training path is `/path/to/dataset1/train/`, `labels.cache` is saved at `/path/to/dataset1/train/`. How will this effect run 1 training if run 2 overwrites labels.cache or vice versa?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @rnguyen-bah, thank you for reaching out and for providing detailed information about your training setup! 🚀 This is an automated response to help you get started—an Ultralytics engineer will also assist you soon.

For questions regarding concurrent training and dataset caching, it's helpful to have a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/)—if you can share a simplified script or reproduction steps, that would be great and will help us debug faster.

We also recommend reviewing the [Docs](https://docs.ultralytics.com/) for guidance on dataset management, [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage, and best practices. If you haven’t already, please ensure the issue persists with the latest version of Ultralytics and all requirements ([pyproject.toml](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml)) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

For real-time help, join our [Discord](https://discord.com/invite/ultralytics) 🎧, or check out [Discourse](https://community.ultralytics.com/) and our [Subreddit](https://reddit.com/r/Ultralytics). You can also run YOLO in verified environments with all dependencies preinstalled:
- **Notebooks:** <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud:** [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon:** [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker:** [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. Our CI verifies all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on multiple platforms every 24 hours and on every commit.

Thanks again for your question! An Ultralytics engineer will review your issue and provide further guidance soon.

**Comment by glenn-jocher** [comment]
Yes, there can be issues with concurrent training when datasets share the same base paths. The cache system uses [hash verification](https://docs.ultralytics.com/reference/data/dataset/#get_labels) to validate cache files, but race conditions can occur when multiple processes try to write to the same `labels.cache` location simultaneously.

In your example, both configs would attempt to cache at `/path/to/dataset1/train/labels.cache`, but with different content due to the dataset combinations. This could cause training instability or cache corruption.

To avoid this, consider organizing your datasets with unique base directories for each training configuration, or use the `cache` parameter to specify different cache locations for each run:

```bash
yolo detect train data=training_config1.yaml model=yolo11n.yaml epochs=100 imgsz=640 cache=cache1
yolo detect train data=training_config2.yaml model=yolo11n.yaml epochs=100 imgsz=640 cache=cache2
```

**Comment by rnguyen-bah** [comment]
Thanks! I'll consider using unique base directories for the training configurations for my use case. Just to double check, if I run multiple concurrent trainings using the same data configuration then that should work correctly? e.g.) I submit 2 training jobs for `training_config1.yaml` but specify different training hyperparameters instead.

I tried specifying different names for the `cache` parameter like you suggested but it doesn't look like it's working. I thought that `cache` should be type `bool` and specifies whether to cache images in directory?

**Comment by glenn-jocher** [comment]
You're absolutely right about the `cache` parameter - I apologize for the confusion. The `cache` parameter is indeed a boolean (or can be set to "ram"/"disk" for different caching modes), not for specifying custom cache file paths.

For concurrent training using the exact same data configuration, that should work correctly since both runs would generate identical cache content. The hash verification system should handle this scenario properly, though there might be brief race conditions during initial cache creation if both jobs start simultaneously.

Your approach of using unique base directories for different dataset combinations is the best solution for avoiding cache conflicts when training with different data configurations.
